src/repositories/updater.rs

```diff
@@ -0,0 +1,366 @@
+    }
+
+    fn store_repository(&self, conn: &mut Client, host: &str, repo: Repository) -> Result<i32> {
+        trace!(
+            "storing {} repository stats for {}",
+            host,
+            repo.name_with_owner,
+        );
+        let data = conn.query_one(
+            "INSERT INTO repositories (
+                 host, host_id, name, description, last_commit, stars, forks, issues, updated_at
+             ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, NOW())
+             ON CONFLICT (host, host_id) DO
+             UPDATE SET
+                 name = $3,
+                 description = $4,
+                 last_commit = $5,
+                 stars = $6,
+                 forks = $7,
+                 issues = $8,
+                 updated_at = NOW()
+             RETURNING id;",
+            &[
+                &host,
+                &repo.id,
+                &repo.name_with_owner,
+                &repo.description,
+                &repo.last_activity_at,
+                &(repo.stars as i32),
+                &(repo.forks as i32),
+                &(repo.issues as i32),
+            ],
+        )?;
+        Ok(data.get(0))
+    }
+
+    fn delete_repository(&self, conn: &mut Client, host_id: &str, host: &str) -> Result<()> {
+        trace!(
+            "removing repository stats for host ID `{}` and host `{}`",
+            host_id,
+            host
+        );
+        conn.execute(
+            "DELETE FROM repositories WHERE host_id = $1 AND host = $2;",
+            &[&host_id, &host],
+        )?;
+        Ok(())
+    }
+}
+
+pub(crate) fn repository_name(url: &str) -> Option<RepositoryName> {
+    static RE: Lazy<Regex> = Lazy::new(|| {
+        Regex::new(r"https?://(?P<host>[^/]+)/(?P<owner>[\w\._/-]+)/(?P<repo>[\w\._-]+)").unwrap()
+    });
+
+    let cap = RE.captures(url)?;
+    let host = cap.name("host").expect("missing group 'host'").as_str();
+    let owner = cap.name("owner").expect("missing group 'owner'").as_str();
+    let repo = cap.name("repo").expect("missing group 'repo'").as_str();
+    Some(RepositoryName {
+        owner,
+        repo: repo.strip_suffix(".git").unwrap_or(repo),
+        host,
+    })
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct RepositoryName<'a> {
+    pub owner: &'a str,
+    pub repo: &'a str,
+    pub host: &'a str,
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::context::Context;
+
+    #[test]
+    fn test_repository_name() {
+        macro_rules! assert_name {
+            ($url:expr => ($owner:expr, $repo:expr, $host:expr)) => {
+                assert_eq!(
+                    repository_name($url),
+                    Some(RepositoryName {
+                        owner: $owner,
+                        repo: $repo,
+                        host: $host,
+                    })
+                );
+            };
+            ($url:expr => None) => {
+                assert_eq!(repository_name($url), None);
+            };
+        }
```

Comment:
This doesn't have to be a macro, it can just be a normal function.

Reply:
True, it's been simplified enough now.

Reply:
Sorry - you changed it to take an `Option`, I was suggesting to remove the `($url:expr => None) => {` bit altogether and just require it to be non-optional; that will make the code a lot less verbose. You can still use `assert_eq` explicitly when you want to test it recognized.